### PR TITLE
Created mapping structure for the NI_XYZ control device

### PR DESCRIPTION
**Pull request by AquaDragonTRS**
(#36 ) Created mapping module for NI_XYZ. Code was written based off the NI_XZ code because the mappings are structurally similar with the inclusion of a few attributes.

Code requires testing.

### CODE REVIEW

**Comment by AquaDragonTRS** [comment]
I've implemented the changes in the checklist. There were some parts in [fauxnixz.py] which I assumed to be arbitrary for the purposes of creating a dataset with numbers (e.g.` data['x'].fill(0.0)`) so I just filled in some numbers.

**Comment by AquaDragonTRS** [comment]
@rocco8773 I have several XYZ files in this folder on midas but I'll list just one:
/data/BAPSF_Data/TDS/19_Apr/06_TDSdipole38_37_XYZ.hdf5

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        if not isinstance(id, h5py.h5g.GroupID):
+            raise ValueError('{} is not a GroupID'.format(id))
+
+        # create control group
+        gid = h5py.h5g.create(id, b'NI_XYZ')
+        h5py.Group.__init__(self, gid)
+
+        # store number on configurations
+        self._knobs = self.Knobs(self, n_motionlists, sn_size)
+
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: improve initialization and clearing of attributes (1 of 3)
```suggestion
        # initialize some attributes
        self._ml = []  # list of motion lists
        self.configs = {}  # configurations dictionary
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        # build NN of each motion list
+        for i in range(self.knobs.n_motionlists):
+            # find divisible numbers
+            sn_div = []
+            for j in range(sn_size_for_ml[i]):
+                if sn_size_for_ml[i] % (j + 1) == 0:
+                    sn_div.append(j + 1)
+
+            # build (Nx, Ny, Nz)
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: the `build (Nx, Ny, Nz)` code block (replaces lines 220-226) does not ensure `Nx * Ny * Nz = number of shots`

```suggestion
            # -- build NN = [(Nx, Ny, Nz), ..,] --
            # define Nx
            sn_div_index = random.randint(0, len(sn_div) - 1)
            Nx = sn_div[sn_div_index]

            # define Ny
            sn_div.clear()
            sn_left = int(sn_size_for_ml[i] / Nx)
            for j in range(sn_left):
                if sn_left % (j + 1) == 0:
                    sn_div.append(j + 1)
            sn_div_index = random.randint(0, len(sn_div) - 1)
            Ny = sn_div[sn_div_index]

            # define Nz
            Nz = int(sn_left / Ny)
            NN.append((Nx, Ny, Nz))
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+                'y0': np.float64(0.0),
+                'Ny': np.float64(NN[i][1]),
+                'dy': np.float64(0.5),
+                'z0': np.float64(0.0),
+                'Nz': np.float64(NN[i][2]),
+                'dz': np.float64(0.5),
+                'fan_XYZ':
+                    np.bytes_('FALSE') if bool(i % 2)
+                    else np.bytes_('TRUE'),
+                'min_zdrive_steps': np.float64(16491),
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: use **NI_XYZ** real value
```suggestion
                'min_zdrive_steps': np.float64(-2000000),
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                'max_zdrive_steps': np.float64(-22833),
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: use **NI_XYZ** real value
```suggestion
                'max_zdrive_steps': np.float64(2000000),
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                'min_ydrive_steps': np.float64(16491),
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: use **NI_XYZ** real value
```suggestion
                'min_ydrive_steps': np.float64(-2469350),
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+                'max_ydrive_steps': np.float64(-22833),
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: use **NI_XYZ** real value
```suggestion
                'max_ydrive_steps': np.float64(2546296),
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+                 for i in range(self.knobs.n_motionlists)]
+            # self._configs[config_name]['motion lists'] = \
+            #     self._motionlist_names
+
+    def _set_xyz_attrs(self):
+        """Set the 'NI_XYZ' group attributes"""
+        self.attrs.update({
+            'Device name': np.bytes_('NI_XYZ'),
+            'Type': np.bytes_('Experimental control'),
+            'Description': np.bytes_(
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: use **NI_XYZ** real values (replaces lines 279-283)

```suggestion
            'Description': np.bytes_(
                '3D Drive from SMPD using National Instruments '
                'ethernet motors (repackaged Applied Motion '
                'Systems drives)'),
            'Module IP address': np.bytes_('192.168.7.76'),
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+                'The NI_XYZ device provides accurate probe position '
+                'control on the central horizontal plane on LAPD.\n'
+                'Uses National Instruments motion hardware.'),
+            'Module IP address': np.bytes_('192.168.7.32'),
+            'Module VI path': np.bytes_('Modules\\NI_XYZ\\NI_XYZ.vi'),
+            'Created data': np.bytes_('1/28/2015 3:03:58 PM'),
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: use **NI_XYZ** real values

```suggestion
            'Created data': np.bytes_('9/9/2018 2:49:51 PM'),
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    def populate(self):
+        """
+        Updates the control group structure (Groups, Datasets, and
+        Attributes)
+        """
+        # clear group before rebuild
+        self.clear()
+
+        # re-initialize key lists
+        self._ml = []
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: improve initialization and clearing of attributes (2 of 3)

```suggestion
        self._ml.clear()
```

File: bapsflib/_hdf/maps/controls/tests/fauxnixyz.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+        # re-initialize key dicts
+        self.configs = {'config01': {}}
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/tests/fauxnixyz.py]: improve initialization and clearing of attributes (3 of 3)

```suggestion
        self.configs.clear()
        self.configs['config01'] = {}
```

**Comment by rocco8773** [comment]
will close #36

File: bapsflib/_hdf/maps/controls/__init__.py [hunk 11]
```diff
@@ -12,11 +12,11 @@
 Package of control device mapping classes and their constructor
 (:class:`~.map_controls.HDFMapControls`).
 """
-from . import (clparse, contype, map_controls, n5700ps, nixz, sixk,
-               templates, waveform)
+__all__ = ['clparse', 'contype', 'ConType', 'HDFMapControls',
+           'map_controls', 'n5700ps', 'nixz', 'nixyz', 'sixk',
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/__init__.py]: alphabetize list

```suggestion
           'map_controls', 'n5700ps', 'nixyz', 'nixz', 'sixk', 
```

File: bapsflib/_hdf/maps/controls/__init__.py [hunk 12]
```diff
@@ -12,11 +12,11 @@
 Package of control device mapping classes and their constructor
 (:class:`~.map_controls.HDFMapControls`).
 """
-from . import (clparse, contype, map_controls, n5700ps, nixz, sixk,
-               templates, waveform)
+__all__ = ['clparse', 'contype', 'ConType', 'HDFMapControls',
+           'map_controls', 'n5700ps', 'nixz', 'nixyz', 'sixk', 
+           'templates', 'waveform']
+
+from . import (clparse, contype, map_controls, n5700ps, nixz, nixyz,
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/__init__.py]: alphabetize imports

```suggestion
from . import (clparse, contype, map_controls, n5700ps, nixyz, nixz, 
```

**Comment by rocco8773** [comment]
Note that the `'r'`, `'theta'`, and `'phi'` fields in the dataset are suppose to represent spherical coordinates of the probe tip with respect to the ball-valve pivot point.  However,  the current calculation of these values do not accurately represent spherical coordinates.  The current calculations go like...

image

I suggest we add a key `'ball-valve pivot distance'` to the mapping `configs` dictionary.  This should then get copied into the `info` dictionary generated by [`HDFReadControl`](https://github.com/BaPSF/bapsflib/blob/b25f05370c0fc601aa7f8e1dff4f17eeec1f8e45/bapsflib/_hdf/utils/hdfreadcontrol.py#L508-L518) during instantiation.

File: bapsflib/_hdf/maps/controls/nixyz.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        # get dataset
+        try:
+            dset = self.group[self.construct_dataset_name()]
+        except KeyError:
+            why = ("Dataset '" + self.construct_dataset_name()
+                   + "' not found")
+            raise HDFMappingError(self.info['group path'], why=why)
+
+        # ---- define general config values                         ----
+        # none exist
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/nixyz.py]: add general items about probe drive pivot (see [comment](https://github.com/BaPSF/bapsflib/pull/39#issuecomment-492894317))

```suggestion
        self.configs[cname].update({
            'Note': "The 'r', 'theta', and 'phi' fields in the "
                    "NI_XYZ data set are suppose to represent "
                    "spherical coordinates of the probe tip with "
                    "respect to the pivot point of the probe drive, "
                    "but the current calculation and population of the"
                    "fields is inaccurate.  For user reference, the "
                    "distance between the probe drive pivot point and"
                    "LaPD axis is (Lpp =) 58.771 cm.",
            'Lpp': 58.771 * u.cm,
        })
```

File: bapsflib/_hdf/maps/controls/nixyz.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        for name in self.group:
+            if isinstance(self.group[name], h5py.Group):
+                _ml_names.append(name)
+
+        # a motion list group must have the attributes
+        # Nx, Ny, Nz, dx, dy, dz, x0, y0, z0
+        names_to_remove = []
+        for name in _ml_names:
+            if all(attr not in self.group[name].attrs
+                   for attr in ('Nx', 'Ny', 'Nz', 'dx', 'dy', 'dz', 'x0', 'y0', 'z0')):
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/nixyz.py]: PEP8 line continuation

```suggestion
                   for attr in ('Nx', 'Ny', 'Nz', 'dx', 'dy', 'dz',
                                'x0', 'y0', 'z0')):
```

File: bapsflib/_hdf/maps/controls/nixyz.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+        # check dset for 'x', 'y' and 'z' fields
+        fx = 'x' not in dset.dtype.names
+        fy = 'y' not in dset.dtype.names
+        fz = 'z' not in dset.dtype.names
+        if fx and fy and fz:
+            why = "Dataset '" + dset.name \
+                  + "' missing fields 'x', 'y' and 'z'"
+            raise HDFMappingError(self.info['group path'], why=why)
+        elif fx or fy or fz:
+            mlist = ('x', fx), ('y', fy), ('z', fz)
```
Comment (rocco8773) [inline bapsflib/_hdf/maps/controls/nixyz.py]: improve readability

```suggestion
            mlist = [('x', fx), ('y', fy), ('z', fz)]
```

**Comment by rocco8773** [comment]
@AquaDragonTRS Looks like all the tests passed.  Good job and thanks for doing this!

**Pull request merged**